app/src/test/java/com/codesoom/assignment/application/TaskServiceTest.java

```diff
@@ -0,0 +1,153 @@
+            @Test
+            @DisplayName("주어진 id에 해당하는 할 일을 리턴한다")
+            void itReturnsValidTask() {
+                Task task = taskService.getTask(1L);
+                assertThat(task.getTitle()).isEqualTo(BEFORE_TASK_TITLE);
+            }
+        }
+
+        @Nested
+        @DisplayName("만약 저장되어 있지 않은 할 일의 id가 주어진다면")
+        class ContextWithInvalidId {
+            @Test
+            @DisplayName("할 일을 찾을 수 없다는 예외를 던진다")
+            void itReturnsErrorMessageException() {
+                assertThatThrownBy(() -> taskService.getTask(100L))
+                        .isInstanceOf(TaskNotFoundException.class);
+            }
+        }
+    }
+
+    @Nested
+    @DisplayName("createTask 메서드는")
+    class Describe_createTask {
+        @Test
+        @DisplayName("title을 입력받아 새로운 할 일을 생성하고 할 일을 리턴한다")
+        void itReturnsNewTask() {
+            int oldIndex = taskService.getTasks().size()-1;
+            Long oldId = taskService.getTasks().get(oldIndex).getId();
+
+            Task newTask = new Task();
+            newTask.setTitle(CREATE_TASK_TITLE);
+
+            taskService.createTask(newTask);
+
+            assertThat(taskService.getTasks().get(oldIndex+1).getId()).isEqualTo(oldId + 1L);
+            assertThat(taskService.getTasks().get(oldIndex+1).getTitle()).isEqualTo(CREATE_TASK_TITLE);
+        }
+    }
+
+    @Nested
+    @DisplayName("updateTask 메서드는")
+    class Describe_updateTask {
+        @Nested
+        @DisplayName("만약 저장되어 있는 할 일의 id가 주어진다면")
+        class ContextWithValidId {
+            @Test
+            @DisplayName("주어진 id에 해당하는 할 일의 title을 수정하고 할 일을 리턴한다")
+            void itReturnsValidUpdatedTask() {
+                Task source = new Task();
+                source.setTitle(UPDATE_TASK_TITLE);
+                taskService.updateTask(1L, source);
+
+                Task task = taskService.getTask(1L);
+                assertThat(task.getTitle()).isEqualTo(UPDATE_TASK_TITLE);
+            }
+        }
+
+        @Nested
+        @DisplayName("만약 저장되어 있지 않은 할 일의 id가 주어진다면")
+        class ContextWithInvalidId {
+            @Test
+            @DisplayName("할 일을 찾을 수 없다는 예외를 던진다")
+            void itReturnsErrorMessageException() {
+                Task source = new Task();
+                source.setTitle(UPDATE_TASK_TITLE);
+
+                assertThatThrownBy(() -> taskService.updateTask(100L, source))
+                        .isInstanceOf(TaskNotFoundException.class);
+            }
+        }
+    }
+
+    @Nested
+    @DisplayName("deleteTask 메서드는")
+    class Describe_deleteTask {
+        @Nested
+        @DisplayName("만약 저장되어 있는 할 일의 id가 주어진다면")
+        class ContextWithValidId {
+            @Test
+            @DisplayName("유효한 id에 해당하는 할 일을 삭제하고 빈 문자열을 리턴한다")
+            void itDeletesTaskAndReturnsEmptyString() {
+                taskService.deleteTask(1L);
+                assertThat(taskService.getTasks().toString()).isEqualTo("[]");
+            }
+        }
```

Comment:
앗 이 테스트가 이상합니다. `deleteTask` 메소드를 검사하는 테스트인데요, `deleteTask` 메소드 실행 결과를 검사하는 것이 아니라 그 실행으로 인해 발생한 사이드 이펙트를 테스트하고 있습니다. 작성하신 테스트 설명대로라면 다음과 같이 작성되어야 합니다.

```suggestion
    @DisplayName("deleteTask 메서드는")
    class Describe_deleteTask {
        @Nested
        @DisplayName("만약 저장되어 있는 할 일의 id가 주어진다면")
        class ContextWithValidId {
            @Test
            @DisplayName("유효한 id에 해당하는 할 일을 삭제하고 빈 문자열을 리턴한다")
            void itDeletesTaskAndReturnsEmptyString() {
                String result = taskService.deleteTask(1L);
                assertThat(result).isEqualTo("");
                // 삭제되었는지도 확인해준다
            }
        }
```

Reply:
알고보니 저의 `delete` 행동은 모두 반환형이 `void`입니다. 그래서 따로 삭제 된 `Task`에 대해서 조회할 수 있는 방법이 없습니다. 코드를 유지시키고, 테스트의 증명과정을 좀 변형시켜봤습니다. 아무래도 가장 만만한게 `size()`변화를 살펴보는 것이고 해당 `givenValidId`이 더이상 유효하지 않음을 보이는 것이라 생각됩니다.

```
@Nested
        @DisplayName("만약 Url에 저장되어 있는 할 일의 id가 주어진다면")
        class ContextWithValidUrlId {
            private final Long givenValidId = 1L;

            @Test
            @DisplayName("주어진 id에 해당하는 할 일을 삭제한다")
            void itReturnsEmptyString() {

                taskController.delete(givenValidId);

                Assertions.assertEquals(taskController.list().size(), 0, "삭제 이후 전체 할 일의 리스트는 사이즈가 0이어야 한다");
                Assertions.assertThrows(TaskNotFoundException.class, () -> taskController.detail(1L),"삭제 이후 주어진 id에 대한 할 일의 조회는 예외를 발생시킨다");
            }
        }
```

Reply:
사이즈 감소가 반드시 삭제 성공을 의미하지 않는다는 점을 고려할 수 있어야 합니다. 한편으로 그것은 내부 구조를 알고 있는 사람의 관점일 수도 있습니다. 테스트 스코프에서 벗어나는 지식일 수 있어요. 삭제하더라도 리스트가 줄지 않는 기능으로 언젠가 업데이트가 되면(그런 일이 일어날 수도 있습니다) 이 테스트도 수정이 되어야겠죠.

여기에서는 `assertThrows`만으로 충분한 것 같습니다. `assertEquals`는 필요 이상인 것 같아요.

Reply:
네 알겠습니다 `size()` 요너석은 앞으로 테스트 작성에서 우선순위에 두지 않아보겠습니다!